File: src/main/java/ting/leetcode/java/arrays101/MinimumLengthOfStringAfterDeletingSimilarEnds.java
package ting.leetcode.java.arrays101;

public class MinimumLengthOfStringAfterDeletingSimilarEnds {
    public int minimumLength(String s) {
        char[] sArray = s.toCharArray();
        int left = 0;
        int right = sArray.length - 1;

        while (left < right) {
            if (sArray[left] != sArray[right]) {
                break;
            }

            char streakChar = sArray[left];
            while (sArray[left] == streakChar && left < right) {
                ++left;
            }
            while (sArray[right] == streakChar && left <= right) {
                --right;
            }
        }
        return right - left + 1;
    }
}

File: kotlin-solution/src/leetcode/arrays101/MinimumLengthOfStringAfterDeletingSimilarEnds.kt
package leetcode.arrays101

class MinimumLengthOfStringAfterDeletingSimilarEnds {
    fun minimumLength(s: String): Int {
        var left = 0
        var right = s.lastIndex

        while (left < right) {
            if (s[left] != s[right]) break
            val streakChar = s[left]

            while (s[left] == streakChar && left < right) {
                ++left
            }

            while (s[right] == streakChar && left <= right) {
                --right
            }
        }

        return right - left + 1
    }
}
